# Burrows-Wheeler Transform

Compression techniques work by finding repeated patterns in the data and encoding the duplications more compactly.
http://en.wikipedia.org/wiki/Burrows%E2%80%93Wheeler_transform

## How to run

1. download bwt.erl
2. launch Erlang shell
3. compile:  c(bwt)
4. run bwt:bwt(String)

## Example
	1> bwt:bwt("BANANA") 
	> Output: BNN$AAA
	[66,78,78,"$",65,65,65]
	2> 


# Smith-Waterman Algorithm 

The Smithâ€“Waterman algorithm performs local sequence alignment; that is, for determining similar regions between two strings or nucleotide or protein sequences.
http://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm

## How to run
1. download sw.erl
2. launch Erlang shell
3. compile: c(sw)
4. run: sw:sw(String1,String2).

## Example
	2> sw:sw("AGGTCA","CGCT").
	GTC
	| |
	G-C
	ok
	3> 

You may also run sw function with random sequences of any length as:

	1> sw:sw(20,40).
	CAGAGAAGCGT-GCG--TGC
	||| |  || | |||  |||
	CAG-G--GCATCGCGTTTGC
	ok
	2> 


# Seeking matches in a fasta file

You may find all matches of your sequence against some reference sequence contained in the .fasta file.
 	
 	1> swp:main(10,"ATGTGACACAGATCACTGCGGCCTTGACCTCCCAGGCTCCAGGTGGTTCTT","21","data/human_g1k_v37_decoy.fasta").

 The first argument means the number of parallel process. The third argument is a name of the reference sequence contained in the fasta file. In this example this is 21st chromosome.
 The function assumes that there is an index file containing the pointers to all reference sequences in the fasta file. You can build this index file using simple python application:

 	>>> python make_index.py data/human_g1k_v37_decoy.fasta
 	>>>

 You need to build an index file only once.
 
	
	
# Big file processing 

These are small tools to help process large files in Erlang.  In general, the strategy is to read in the file as an array of possibly overlapping Erlang binary "chunks".  These can then be processed in parallel/concurrently.

## How to run
1. download bio_pfile.erl
2. launch Erlang shell
3. compile: c(bio_pfil)
4. run: bio_pfile:read(Filename,NumerOfChunks) or bio_pfile:read(FileName,NumberOfChunks,SizeOfOverlap)
        both of which return an array of chunk elements: {{StartPos,Length},BinaryData}

## Example

    1> Data = bio_pfile:read("../data/GCA_000001405.15_GRCh38_full_analysis_set.fna",10000).
    [{{0,32553715},<<">chr1  AC:CM000663.2  gi:568336023  LN:248956422  rl:Chromosome  M5:6aef897c3d6ff0c78aff06ac189178dd     AS"...>>},
    {{32543715,32563715},<<"ACCTCATAGATTGGTCATCTTTTTCTC\nCTATATTTCTCTAATATTTAATCTCTCTCTCTCTCTCTCTTTGTATGTGCATTGCCTTTGGAGAGATTTC\nC"...>>},
    {{65097430,32563715},<<"AATCAAGAAAATATGTTTACCAAAA\nTGCATTGCAATTTTCCCAAACCTGAGTCTTCAAATAACAAACATGAACTTATAGGTACTGTGAACTAGAA"...>>},
    {{97651145,32563715},<<"CAAGAATTGAGGTTTGGGAAACT\nCCATCTAGATTTCAGAGGATGTATGGAAATACCTGGATGTCCAGGCAGTAGTTTGCTGCAAGGGTGTG"...>>},
    {{813832875,32563715},<<"TT\nT"...>>},
    {{846386590,...},<<...>>},
    {{...},...},
    {...}|...]
    2> length(Data).                                                                        
    10001
    3> lists:nth(1,Data).                                                                   
    {{0,32553715},<<">chr1  AC:CM000663.2  gi:568336023  LN:248956422  rl:Chromosome  M5:6aef897c3d6ff0c78aff06ac189178dd  AS:GRC"...>>}
    4>  


5. run: bio_pfile:spawn_find_pattern(ChunkArray,BinaryPattern) which returns an array of all the stat positions where the pattern was found as {StartPosition,LengthOfPattern}.

## Example

    4>  bio_pfile:spawn_find_pattern(Data,<<"TATATTCAGTCTTTCTAACACCATTTATTGAAGAGACTGTAG">>).
    [{162758595,42}]
